File: day4/day4.cpp
#include <fstream>
#include <sstream>
#include <vector>
#include <fmt/format.h>

using namespace std;

class Board
{
public:
	Board()
	{
		for (size_t i = 0; i < 25; i++)
		{
			numbers[i] = 0;
			marked[i] = false;
		}
		won = false;
	}

	bool mark_and_check(int num)
	{
		size_t pos;
		for (pos = 0; pos < 25; pos++)
		{
			if (numbers[pos] == num)
			{
				break;
			}
		}
		if (pos < 25)
		{
			marked[pos] = true;
			won = won || check_row(pos) || check_col(pos);
		}
		return won;
	}

	int score(int num) const
	{
		int score = 0;
		for (size_t i = 0; i < 25; i++)
		{
			if (!marked[i])
			{
				score += numbers[i];
			}
		}
		return num * score;
	}

private:
	bool check_row(size_t pos) const
	{
		pos -= pos % 5;
		for (size_t end = pos + 5; pos < end; pos++)
		{
			if (!marked[pos])
			{
				return false;
			}
		}
		return true;
	}

	bool check_col(size_t pos) const
	{
		for (pos %= 5; pos < 25; pos += 5)
		{
			if (!marked[pos])
			{
				return false;
			}
		}
		return true;
	}

	int numbers[25];
	bool marked[25];
public:
	bool won;

	friend istream& operator>>(istream&, Board& b);
};

istream& operator>>(istream& in, Board& b)
{
	for (size_t i = 0; i < 25; i++)
	{
		if (!(in >> b.numbers[i]))
		{
			break;
		}
	}
	return in;
}

istream& operator>>(istream& in, vector<int>& numbers)
{
	string line;
	if (!getline(in, line))
	{
		return in;
	}

	istringstream is(line);
	string num;
	while (getline(is, num, ','))
	{
		numbers.push_back(stoi(num));
	}

	return in;
}

static int part1(vector<int>& numbers, vector<Board>& boards)
{
	for (auto n = numbers.begin();
	     n != numbers.end();
	     n = numbers.erase(n))
	{
		for (auto& b: boards)
		{
			if (b.mark_and_check(*n))
			{
				return b.score(*n);
			}
		}
	}
	return 0;
}

static int part2(vector<int>& numbers, vector<Board>& boards)
{
	for (auto n: numbers)
	{
		for (auto i = boards.begin(); i != boards.end(); )
		{
			if (!i->mark_and_check(n))
			{
				++i;
			}
			else if (boards.size() != 1)
			{
				i = boards.erase(i);
			}
			else
			{
				return i->score(n);
			}
		}
	}
	return 0;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<int> numbers;
	in >> numbers;

	vector<Board> boards;
	Board b;
	while (in >> b)
	{
		boards.emplace_back(b);
	}
	in.close();

	fmt::print("Part1: {}\n", part1(numbers, boards));
	fmt::print("Part2: {}\n", part2(numbers, boards));
	return 0;
}

File: day4/day4.py
#!/usr/bin/env python3

class Board(object):
    def __init__(self, numbers):
        assert len(numbers) == 25, "Boards have exactly 25 numbers"
        self.numbers = numbers
        self.marked = [False] * 25
        self.won = False

    def mark_and_check(self, number):
        try:
            pos = self.numbers.index(number)
            self.marked[pos] = True

            # NOTE: this will short-circuit the expensive checks
            # ordered by their cost
            self.won = self.won \
                or self.check_row(pos) \
                or self.check_column(pos)

        except ValueError:
            pass

        return self.won

    def check_row(self, pos):
        start = pos - (pos%5)
        for x in range(start, start+5):
            if not self.marked[x]:
                return False

        return True

    def check_column(self, pos):
        for y in range(pos%5, 25, 5):
            if not self.marked[y]:
                return False

        return True

    def score(self, number):
        score = 0
        for i, x in enumerate(self.numbers):
            if not self.marked[i]:
                score += x

        return score * number

def parse_boards(txt):
    numbers = []
    boards = []
    board = []
    state = 0
    for line in txt.splitlines():
        if state == 0:
            # read numbers
            if line:
                numbers = list(int(x) for x in line.split(","))
            else:
                state = 1
        else:
            # read boards
            if line:
                board += list(int(x) for x in line.split())
            else:
                boards.append(Board(board))
                board = []
    if board:
        boards.append(Board(board))

    return numbers, boards

def part1(boards, numbers):
    for n in numbers:
        for b in boards:
            if b.mark_and_check(n):
                return b.score(n)
    return 0

def part2(boards, numbers):
    for n in numbers:
        for b in boards:
            b.mark_and_check(n)
        if len(boards) == 1 and boards[0].won:
            return boards[0].score(n)
        else:
            boards = list(filter(lambda b: not b.won, boards))
    return 0

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        numbers, boards = parse_boards(txt)
        print("Part1:", part1(boards, numbers))
        print("Part2:", part2(boards, numbers))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
